### PR TITLE
Add config option and compiler support to reuse across OS's

### PR DESCRIPTION
**Pull request by psakievich**
This PR enables the compilers to be reused across compatible os's and adds a configuration option for users to set this to their needs.

### CODE REVIEW

File: lib/spack/spack/solver/concretize.lp [hunk 1]
```diff
@@ -1186,7 +1186,8 @@ error(100, "{0} compiler '%{1}@{2}' incompatible with 'os={3}'", Package, Compil
      node_compiler(node(X, Package), CompilerID),
      compiler_name(CompilerID, Compiler),
      compiler_version(CompilerID, Version),
-     not compiler_os(CompilerID, OS),
+     compiler_os(CompilerID, CompilerOS),
+     not os_compatible(CompilerOS, OS)
```
Comment (alalazo) [inline lib/spack/spack/solver/concretize.lp]: There's a syntax error:
```suggestion
     not os_compatible(CompilerOS, OS),
```

Comment (psakievich) [reply lib/spack/spack/solver/concretize.lp]: Thanks! I copy pasted this into the web IDE. 🤦‍♂️

**Comment by psakievich** [comment]
Need to add for user added os compatibility.
```
concretizer:
  os_compatible:
  - [rhel8, rhel7]
```

File: etc/spack/defaults/concretizer.yaml [hunk 2]
```diff
@@ -42,3 +42,8 @@ concretizer:
     # "minimal": allows the duplication of 'build-tools' nodes only (e.g. py-setuptools, cmake etc.)
     # "full" (experimental): allows separation of the entire build-tool stack (e.g. the entire "cmake" subDAG)
     strategy: minimal
+  # Option to specify compatiblity between operating systems for reuse of compilers and packages
+  # Specified as a key: [list] where the key is the os that is able to reuse, and the list contains the OS's 
+  # it can reuse. Note this is a directional compatibility so mutual compatibility between two OS's 
+  # requires two entries i.e. os_compatible: {sonomar:[monterey], monterey:[sonoma]}
```
Comment (becker33) [inline etc/spack/defaults/concretizer.yaml]:
```suggestion
  # requires two entries i.e. os_compatible: {sonoma: [monterey], monterey: [sonoma]}
```

File: etc/spack/defaults/concretizer.yaml [hunk 3]
```diff
@@ -42,3 +42,8 @@ concretizer:
     # "minimal": allows the duplication of 'build-tools' nodes only (e.g. py-setuptools, cmake etc.)
     # "full" (experimental): allows separation of the entire build-tool stack (e.g. the entire "cmake" subDAG)
     strategy: minimal
+  # Option to specify compatiblity between operating systems for reuse of compilers and packages
+  # Specified as a key: [list] where the key is the os that is able to reuse, and the list contains the OS's
```
Comment (becker33) [inline etc/spack/defaults/concretizer.yaml]:
```suggestion
  # Specified as a key: [list] where the key is the os that is being targeted, and the list contains the OS's 
```

**Review by becker33 — CHANGES_REQUESTED** [review]
Code all looks good, couple clarifications on the comments.

**Review by becker33 — APPROVED** [review]

**Comment by alalazo** [comment]
This PR didn't remove `os_compatibility.lp`, was that done on purpose? Right now it's not clear to me why we hardcode some compatibility rules rather than having a default configuration.

**Comment by psakievich** [comment]
@alalazo the config was to allow users to extend/customize. I was thinking we would have the "official" and "unofficial/user" entry points.  But we could definitely remove the `*.lp` if that is preferred.

One consequence is that when we coalesce configs this falls under the same issues as multivalued variants and compiler flags.

**Comment by alalazo** [comment]
> I was thinking we would have the "official" and "unofficial/user" entry points.

I assume the "official" one is the config, right? I just find a bit puzzling that we have two ways of doing the same thing. I hope we can get rid of OS logic in the solver in https://github.com/spack/spack/pull/43190 anyhow

> One consequence is that when we coalesce configs this falls under the same issues as multivalued variants and compiler flags.

Not sure I get what you mean here.

**Comment by psakievich** [comment]
I mean if we set the default `os_compatibility` in the default `concretizer.yaml` and then a user were to set a single one, then I think all the defaults would not be in scope for the aggregate config that they were to use.  They would only have the os compatibilities they specified.

My thought was the official would NOT be the config, but rather the ones that are hard coded into the concretizer, and the config just gave users a back door to extend that.

However, 🤔 maybe I'm wrong and it would only wipe out defaults for an os they configured. I will test it out. Happy to do a follow on PR to remove the hard coded ASP file if that is what is best.

**Pull request merged**